File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.java
class Solution {
    /*
        1. b - a가 증가추세이면 a를 고정시키고 b를 계속 증가시켜주면서 수익의 극대화를 구함
        2. 증가추세가 꺽이면 b - a를 sum += 기존 최고 수익, 기존 최고 수익 = 0, a = curr, 1번 실행
    */
    
    public int maxProfit(int[] prices) {        
        int l = 0, r = 1;
        int maxProfit = 0;

        while(r < prices.length){
            if(prices[r - 1] >= prices[r]){ // sell
                maxProfit += prices[r - 1] - prices[l];
                l = r;
            }
            r++;
        }

        if(prices[r - 1] - prices[l] > 0) maxProfit += prices[r - 1] - prices[l];
        
        return maxProfit;
    }
}

File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        """
        Algorithm & DS: Greedy
        Time: O(N), Space: O(1)
        """
        maxProfit = 0

        for i in range(1, len(prices)):
            if prices[i - 1] < prices[i]:
                maxProfit += prices[i] - prices[i - 1]

        return maxProfit
        